File: AI-BigData班第36期/MongoDB/mongoDB_eb102/eb102/mr1.js

var db = db.getSisterDB("eb102");

var showCursorItems = function(cursor){
	while (cursor.hasNext()) {
   		printjson(cursor.next());
	}
}

var findAllAndShow = function(coll_name){
	print('call findAll');
	var cursor = db[coll_name].find();
    showCursorItems(cursor);
}

var findOneAndShow = function(coll_name){
	printjson(db[coll_name].findOne());
}

db.mr1.drop();

db.mr1.insert(
 	[   
		{x:10,y:23,yyy:3455},
		{y:10,z:33},
		{z:10,w:10,x:678},
		{z:10,yyy:"hello",ccc: 1},
		{x:10,w:20,z:10,yyy:"hello"}
	]
);

var mapxx = function(){
	for(var key in this){
		if(key !== '_id'){
			emit(key,{count:1});
		}
	}
}
var reducexx = function(key,emits){
	total = 0;
	for(var i in emits){
		total+=emits[i].count;
	}
	return {count:total};
}

var mrResult = db.runCommand({'mapreduce':'mr1','map':mapxx,'reduce':reducexx,"out":{inline:1}});
printjson(mrResult);

// var mrResult = db.mr1.mapReduce(mapxx,reducexx,{out:'a20200225'});
// findAllAndShow('a20200225');




// var mapxx = function(){
// 	emit(this["age"],{count:1});
// }
// var reducexx = function(key,emits){
// 	total = 0;
// 	for(var i in emits){
// 		total+=emits[i].count;
// 	}
// 	return {count:total};
// }

// var mrResult = db.runCommand({'mapreduce':'usersNonIndex','map':mapxx,'reduce':reducexx,"out":{inline:1}});
// printjson(mrResult);

File: AI-BigData班第36期/MongoDB/mongoDB_eb102/eb102/mr1.py
from bson.code import Code
import pymongo
db = pymongo.MongoClient().eb102

# mapper = Code("""function(){
# 	for(var key in this){
# 		if(key !== '_id'){
# 			emit(key,{'count':1});
# 		}
# 	}
# }
# """)

mapper = Code("function(){emit(this['age'],{count:1});}")

reducer = Code(""" 
	function(key,emits){
	total = 0;
	for(var i in emits){
		total+=emits[i].count;
	}
	return {'count':total};
}
""")

result = db.usersNonIndex.map_reduce(mapper, reducer, "mr1_result_20200225")

for row in result.find():
    print(row)

